### PR TITLE
Initial Feature Election algorithm version for Nvidia FLARE

### PR DESCRIPTION
**Pull request by christofilojohn**
Feature Election for NVIDIA FLARE
### Description
A plug-and-play horizontal federated feature selection framework for tabular datasets in NVIDIA FLARE.

This work originates from FLASH: A framework for Federated Learning with Attribute Selection and Hyperparameter optimization framework a work presented in [FLTA IEEE 2025](https://flta-conference.org/flta-2025/) achieving the best student paper award.

Feature election enables multiple clients with tabular datasets to collaboratively identify the most relevant features without sharing raw data. It works by using conventional Feature selection algorithms in the client side (executor) and performing a weighted aggregation of their results on the server side (controller).

FLASH is available on [Github](https://github.com/parasecurity/FLASH)

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [x] New tests added to cover the changes.
- [x] Quick tests passed locally by running `./runtest.sh`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated.

### CODE REVIEW

File: nvflare/app_opt/feature_election/feature_election.py [hunk 1]
```diff
@@ -0,0 +1,767 @@
...
+            controller.freedom_degree = best_fd
+            logger.info(f"Auto-tuned freedom_degree: {best_fd:.2f} (score: {best_score:.3f})")
+        
+        # Aggregate selections
+        self.global_mask = controller._aggregate_selections(client_selections)
+        
+        # Calculate intersection and union for stats
+        masks = np.array([sel["selected_features"] for sel in client_selections.values()])
+        intersection_mask = np.all(masks == 1, axis=0)
+        union_mask = np.any(masks == 1, axis=0)
```
[the remark this answers is not in the document]
Comment (christofilojohn) [reply nvflare/app_opt/feature_election/feature_election.py]: Horizontal federated learning paradigm makes this assumption by design

File: nvflare/app_opt/feature_election/controller.py [hunk 2]
```diff
@@ -0,0 +1,356 @@
...
+        n_additional = int(np.ceil(np.sum(difference_mask) * self.freedom_degree))
+
+        if n_additional > 0:
+            diff_indices = np.where(difference_mask)[0]
+            diff_scores = aggregated_scores[difference_mask]
+
+            if len(diff_scores) > 0:
+                # Get indices of top scoring features
+                top_indices = np.argpartition(diff_scores, -min(n_additional, len(diff_scores)))
+                top_indices = top_indices[-min(n_additional, len(diff_scores)):]
```
[the remark this answers is not in the document]
Comment (christofilojohn) [reply nvflare/app_opt/feature_election/controller.py]: Fixed in a recent commit

**Comment by christofilojohn** [comment]
it's ready for review as i fixed the licensing issue. Will maintain this code with more, informative examples

File: examples/advanced/feature_election/flare_deployment.py [hunk 3]
```diff
@@ -0,0 +1,350 @@
...
+    
+    # Simulated result retrieval
+    # In production:
+    # from nvflare.fuel.flare_api.flare_api import new_secure_session
+    # session = new_secure_session()
+    # job_result = session.get_job_result(job_id)
+    # global_mask = job_result['global_feature_mask']
+    
+    # For this example, we'll simulate with saved results
+    from nvflare.app_opt.feature_election import load_election_results
```
[the remark this answers is not in the document]
Comment (christofilojohn) [reply examples/advanced/feature_election/flare_deployment.py:146]: load_election_results method is part of the Feature Election module

File: nvflare/app_opt/feature_election/controller.py [hunk 4]
```diff
@@ -0,0 +1,367 @@
...
+        self,
+        freedom_degree: float = 0.1,
+        aggregation_mode: str = 'weighted',
+        min_clients: int = 2,
+        num_rounds: int = 1,
+        task_name: str = "feature_election",
+        aggregator_id: str = "aggregator",
+        persistor_id: str = "persistor",
+        shareable_generator_id: str = "shareable_generator",
+        train_timeout: int = 0
```
[the remark this answers is not in the document]
Comment (christofilojohn) [reply nvflare/app_opt/feature_election/controller.py]: Unused, will remove

File: examples/advanced/feature_election/flare_deployment.py [hunk 5]
```diff
@@ -0,0 +1,350 @@
...
+    
+    # Initialize Feature Election with your parameters
+    fe = FeatureElection(
+        freedom_degree=0.5,  # Will select features between intersection and union
+        fs_method='lasso',   # Feature selection method
+        aggregation_mode='weighted'  # Weight by sample count
+    )
+    
+    # Generate FLARE job configuration
+    job_paths = fe.create_flare_job(
```
[the remark this answers is not in the document]
Comment (christofilojohn) [reply examples/advanced/feature_election/flare_deployment.py:44]: Defined in Feature Election

File: nvflare/app_opt/feature_election/executor.py [hunk 6]
```diff
@@ -0,0 +1,567 @@
...
+        self.global_feature_mask = np.array(global_mask, dtype=bool)
+
+        # Log results
+        logger.info(f"Received global mask: {np.sum(self.global_feature_mask)} features selected")
+
+        # Apply mask to training data if needed
+        if self.X_train is not None:
+            self.X_train = self.X_train[:, self.global_feature_mask]
+            if self.X_val is not None:
+                self.X_val = self.X_val[:, self.global_feature_mask]
```
[the remark this answers is not in the document]
Comment (christofilojohn) [reply nvflare/app_opt/feature_election/executor.py:498]: Comments explain the reasoning, the feature mask is applied after the election

File: nvflare/app_opt/feature_election/tests/test_feature_election.py [hunk 7]
```diff
@@ -0,0 +1,408 @@
...
+        assert 0 <= fe.freedom_degree <= 1
+        assert 'freedom_degree' in stats
+
+    @pytest.mark.skipif(
+        not PYIMPETUS_AVAILABLE,
+        reason="PyImpetus dependency not installed."
+    )
+    def test_freedom_degree_intersection(self, sample_data):
+        """Test freedom_degree=0 gives intersection"""
+        fe = FeatureElection(freedom_degree=0.0, fs_method='pyimpetus')
```
[the remark this answers is not in the document]
Comment (christofilojohn) [reply nvflare/app_opt/feature_election/tests/test_feature_election.py]: pyimpetus is correct

File: examples/advanced/feature_election/basic_usage.py [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+        aggregation_mode='weighted'
+    )
+    
+    # Prepare data splits
+    client_data = fe.prepare_data_splits(
+        df=df,
+        target_col='target',
+        num_clients=5,
+        split_strategy='stratified'
+    )
```
[the remark this answers is not in the document]
Comment (christofilojohn) [reply examples/advanced/feature_election/basic_usage.py]: It is implemented

File: nvflare/app_opt/feature_election/tests/test_feature_election.py [hunk 9]
```diff
@@ -0,0 +1,408 @@
...
+        return df
+    
+    def test_quick_election_basic(self, sample_data):
+        """Test basic quick election"""
+        mask, stats = quick_election(
+            sample_data,
+            target_col='target',
+            num_clients=3,
+            fs_method='lasso'
+        )
```
[the remark this answers is not in the document]
Comment (christofilojohn) [reply nvflare/app_opt/feature_election/tests/test_feature_election.py]: It will be implemented when it's merged. Now it's just a file, not part of flare project

File: examples/advanced/feature_election/requirements.txt [hunk 10]
```diff
@@ -0,0 +1,2 @@
+scikit-learn>=1.0.0
+PyImpetus>=0.0.6
```
[the remark this answers is not in the document]
Comment (christofilojohn) [reply examples/advanced/feature_election/requirements.txt]: Yes, it's utilized in the examples

**Comment by christofilojohn** [comment]
Hello @chester-chen,

Gentle follow-up ping on this refreshed & ready Feature Election contribution (the cleaned-up version to the earlier attempt).

- All unit tests green  
- Greptile silent
- New tests folder structure, similar to other app_opt examples
- Boolean mask handling now robust  
- CodeQL is the last thing running

Ready for final review + merge whenever you have a moment.

Thank you,
Kind Regards,
Ioannis

**Comment by christofilojohn** [comment]
Hi @chesterxgchen, @holgerroth,

This PR is ready for review and from my side—all local tests pass (I do not have access to run CI, but the initial failures were happening because of missing open-source licence headers).

Could you please assign two reviewers, or approve a CI test when you have a moment?

Thank you for your time and help!

**Comment by chesterxgchen** [comment]
@christofilojohn  I have simple question for you -- basically try to understand what is this Algorithm intended for. 

Base on the linked github repo for FLASH  [Github](https://github.com/parasecurity/FLASH), the algorithm is a way to selecting which features to be included in training in the Federated Settings. 

If all sites has the same features ( intersect), then it is horizontal learning, if you select a union, then it is vertical FL.  So what this Algorithm actually do  (key ideas in simple terms, without much details) ?  Why do we need it ?

**Comment by christofilojohn** [comment]
@chesterxgchen  This algorithm is useful for communication efficiency, while maintaining or improving performance in Horizontal Federated Learning settings (no Vertical FL testing or setup has been performed). It's a horizontal feature selection method, that uses feature selection algorithms made for conventional Machine Learning, on each client and aggregates the (feature selection) results - meaning the selected feature sets. 
 We have a parameter - 'freedom degree' that manages the voting process and decides on a feature set between the union and intersection of votes. Essentially, in our paper we used this method as pre-processing, before performing aggregation rounds and saw that we can get about 50% model parameter size reduction on average (while maintaining performance), leading in reduced communication cost - as models that are trained on less features are more compact (tested on sklearn). In noisy datasets, we also got improved f1 score prediction performance.
 This is an Horizontal FL pre-processing method that mainly aims for communication efficiency, with it's main innovation being the utilization of conventional Feature Selection methods in a federated environment.
 Thank you for taking the time to check this, let me know if there are any questions,
Ioannis Christofilogiannis

**Comment by chesterxgchen** [comment]
> We have a parameter - 'freedom degree' that manages the voting process and decides on a feature set between the union and intersection of votes

Thanks. This is very helpful. One more follow-up question.  "who" is voting for the features ? since the freedom degree is chosen by "voting" and what criteria to decide the vote ?  I guess that determines why choose this feature vs. another features --> feature importance.

**Comment by christofilojohn** [comment]
Each client (executor) has a different dataset and distribution due to the horizontal partitioning and so performing the same feature selection method (like Lasso or PyImpetus) will lead to different results. Each client selects some features and gives  a preference score to all features, based on the utilized feature selection method. The server (executor) performs the aggregation of the selection vectors (binary choice and preference) based on the selected freedom degree value. The attached visual explains the process:
https://drive.google.com/file/d/1c_zUoVKFqI9yg9Bl-yX2fnebnVYyFRis/view?usp=sharing

In our paper, the freedom degree value is set by the user and it was acknowledged as a limitation of the work. However, in this pull request, a simple automatic optimization method where multiple values are tested for the freedom degree, has been implemented, but this part is still open research.
Thank you

**Comment by chesterxgchen** [comment]
@christofilojohn  Ok. Assume the the researcher set the degree of freedom, such as 0.4, then each site will choose based on the that degree of freedom to select the features ( according certain local selection algorithms) and then send the server to aggregation. This make sense. but the graph actually confuse me, it confused me from the original github plot. 

if the different degree of freedom causing the feature selected from different intersection of client 1 and 2, but not client 3, we are back to Vertical FL case. But we are intended to targeting Horizontal FL case,  this is seems to contradiction to the goal .

**Comment by christofilojohn** [comment]
Each site will choose its features independently from the freedom degree, it's only in the server (executor) aggregation that such a mechanism is necessary, to make a decision based on the client's results (think of each small circle as the client's preferences and the venn-like diagram as all the preferences that the server sees). The freedom degree tells the server how to handle these client vectors, with 0.1 meaning a more strict approach, that limits the total number of features, for example. Features at the Intersection set are always selected. So it's just a way to control the aggregation behavior, that doesn't influence each client's decision. After the election a global mask is decided and all clients use it.

 Each client performs feature selection locally, with each possible selector having different hyperparameters (PyImpetus being relatively plug and play). For example Lasso has a hyperparameter a (or C, depending on literature) which changes it's feature selection behavior.

 This approach allows existing ML pipelines with feature selection to migrate to horizontal FL, a simple pre-processing Feature selection round that reduces model parameter size for future fl rounds and a way to make any conventional feature selection method (that give us preference scores) to work in FL.

**Comment by chesterxgchen** [comment]
> Each site will choose its features independently from the freedom degree, it's only in the server (executor) aggregation that such a mechanism is necessary, to make a decision based on the client's results (think of each small circle as the client's preferences and the venn-like diagram as all the preferences that the server sees). The freedom degree tells the server how to handle these client vectors, with 0.1 meaning a more strict approach, that limits the total number of features, for example. Features at the Intersection set are always selected. So it's just a way to control the aggregation behavior, that doesn't influence each client's decision. After the election a global mask is decided and all clients use it.
> 
> Each client performs feature selection locally, with each possible selector having different hyperparameters (PyImpetus being relatively plug and play). For example Lasso has a hyperparameter a (or C, depending on literature) which changes it's feature selection behavior.
> 
> This approach allows existing ML pipelines with feature selection to migrate to horizontal FL, a simple pre-processing Feature selection round that reduces model parameter size for future fl rounds and a way to make any conventional feature selection method (that give us preference scores) to work in FL.

Just make me understand better. the degree of freedom is used for Server side for aggregation. no freedom (= 0) is pure intersection, more freedom ( > 0) could produce some union set.  But each site are free to choose what features based on the local section algorithm. 

What if we have 5 same features F1 to F5 on each site ( horizontal FL).  After feature section, site-1 select F1 to F4 and Site-2 selected F2 to F5. The degree of freedom will be > 0.  the aggregated one is Now F1, F2, F3, F4, F5.  the FL Model will be trained on the F1 to F5 even the selection on site-2 is only F2 to F5. is this correct understanding ?

**Comment by christofilojohn** [comment]
Yes, that's correct. 0 Is intersection, 1 is union and the in between values define a set between intersection and union. For example 0.2 means intersection +20% of the features that are in the union but not in the intersection (the freedom degree defines the number of features to be selected from the Union-Intersection set).
  However, in your example F2-F4 are immediately selected (intersection set) but F1 and F5 compete for a spot. With a freedom degree of >0.5 both are selected, due to the np.ceil function (Total Union-Intersection set contains 2 features, fd=0.5 means 50% of that so 1 feature). 
 If freedom degree <= 0.5, then only 1 feature out of the 2 candidates is selected, based a the preference score. Each client gives preference scores for all features (selected or not), so these collected preference scores are aggregated in the server (like a voting system). Based on this process the server has a clear idea of which feature is preferred by all of the clients and selects for example:

F1,F2,F3,F4 - freedom degree = 0.5, when F1 is preferred to F5
F1,F2,F3,F4,F5 - freedom degree = 0.2, in either case

 This is way for the preference for each feature from each client to count, even if the client hasn't selected the specific feature.

Thank you,
Ioannis

**Comment by chesterxgchen** [comment]
Thanks for the explanation, I understand what this PR about now. Maybe Holger and Ziyue can help review this

File: examples/advanced/feature_election/basic_usage.py [hunk 11]
```diff
@@ -0,0 +1,234 @@
+# Copyright (c) 2025, NVIDIA CORPORATION.  All rights reserved.
```
Comment (chesterxgchen) [inline examples/advanced/feature_election/basic_usage.py:1]: We are in the process of moving to a standard structure with our examples 

Basically we don't want to name the examples in random filenames which make it harder for the users to understand the code structures ( as it varies with every example)

the prefer the structures are the followings

client.py
server.py -- optional
job.py  
model.py  -- optional 
download_data.py --> optional 
prepare_data.py -- optional 

the client.py usually associate with client side logics ( client executor code or client training scripts) 
the server.py usually associate with server side logics ( controller, workflow or FL algo) 
the model.py usually model , in your case, there is likely none
the job.py is the code that piece server and client configuration togather ( with JobRecipe or Job API etc) 


If you have multiple jobs/examples, don't mix them up. it will make it hard for user.

**Comment by holgerroth** [comment]
Thanks for your contribution. It might be better to move this example and new components under https://github.com/NVIDIA/NVFlare/tree/main/research

File: examples/advanced/feature_election/basic_usage.py (hunk 11 above)
Comment (christofilojohn) [reply examples/advanced/feature_election/basic_usage.py:1]: Thanks for the feedback, I will update and set as resolved.

**Comment by christofilojohn** [comment]
Thank you for the feedback @holgerroth. The only reason I put it in the app_opt is that I consider it a valid plugin for some workflows that require efficiency. It was not put in research examples as this is not yet a recreation of our paper's experiments (that were done in FLASH framework). This uploaded version performs some experiments on synthetic datasets as proof of concept and is offered as an option for users, who want tabular, horizontal FL with NVFLARE.

> Thanks for your contribution. It might be better to move this example and new components under https://github.com/NVIDIA/NVFlare/tree/main/research

File: nvflare/app_opt/feature_election/tests/test_feature_election.py
Comment (holgerroth) [inline nvflare/app_opt/feature_election/tests/test_feature_election.py:1]: Test should be here so the can run in our CI https://github.com/NVIDIA/NVFlare/tree/main/tests. @YuanTingHsieh for viz

Comment (christofilojohn) [reply nvflare/app_opt/feature_election/tests/test_feature_election.py:1]: Cool, will move it soon, currently working on some changes which I need to test locally before pushing.
Thanks

**Comment by christofilojohn** [comment]
I will be pushing some substantial changes with a revised auto tuning method of the variable and structure that follows our research paper (but with synthetic datasets). @holgerroth Should I put my test file in: tests/unit_test/app_opt/feature_election as test.py and only there? Also will the CI detect it or should I change some configuration files? 
Thank you,
Ioannis

**Comment by chesterxgchen** [comment]
> ests/unit_test/app_opt/feature_election

**Comment by chesterxgchen** [comment]
> I will be pushing some substantial changes with a revised auto tuning method of the variable and structure that follows our research paper (but with synthetic datasets). 
@holgerroth Should I put my test file in: tests/unit_test/app_opt/feature_election as test.py and only there? Also will the CI detect it or should I change some configuration files? Thank you, Ioannis


the code is in app_opt module, yes, the unit test file should be in that location. For CI test,  it should be under tests/integration_tests. 

But we are considering to making big changes to the integration tests. unless the CI test is significant different from unit tests, you might skip that for now.

**Comment by christofilojohn** [comment]
Ok I understand, for now it's just unit tests, that pass locally, I have made changes to the examples, putting one that is simpler and. configurable based on your earlier comments. I am making changes to the documentation and I can push within an hour.
Thank you,
Ioannis
